gcc/rust/expand/rust-macro-expand.h

```diff
@@ -321,12 +321,28 @@ struct MacroExpander
   AST::ASTFragment take_expanded_fragment (AST::ASTVisitor &vis)
   {
     AST::ASTFragment old_fragment = std::move (expanded_fragment);
+    auto accumulator = std::vector<AST::SingleASTNode> ();
     expanded_fragment = AST::ASTFragment::create_error ();
 
     for (auto &node : old_fragment.get_nodes ())
       {
 	expansion_depth++;
 	node.accept_vis (vis);
+	// we'll decide the next move according to the outcome of the macro
+	// expansion
+	if (expanded_fragment.is_error ())
+	  accumulator.push_back (node); // if expansion fails, there might be a
+					// non-macro expression we need to keep
```

Comment:
Do you have any test cases for this or any examples? I have a hard time figuring out what this achieves (sorry, this is on me, not on your code) and why we'd still want to push to the accumulator when we encounter an error.

Reply:
> Do you have any test cases for this or any examples? I have a hard time figuring out what this achieves (sorry, this is on me, not on your code) and why we'd still want to push to the accumulator when we encounter an error.

This block is what fixed the issue.

When recursively expanding on a macro, the expansion may produce nodes that can't be expanded further (macro transcribed completely to a regular Rust expression). However, they are still part of the expansion due to how `take_expanded_fragment` is invoked in most places.

The explanations above may seem a bit dry, so let's see an example, shall we?

Okay, let's see the simplified version of the test case included in this pull request:

```rust
stmt!(
        let a = 1,
        let b = 2,
    );
```

One the first pass of GCC-rs expansion, this expands to a single fragment:

```rust
{ let a = 1; stmt!(let b = 2,) }
```

In this expanded fragment, there are 2 (two) nodes, `let a = 1` and `stmt!(let b = 2,)`.

What will happen to the first node when we expand this _fragment_ again?

... ...

Hmm... The first node will return an expansion error after calling `accept_vis (vis)` on it since it does not contain a macro expression. So, in this case, if we discard the error, what will happen then...?

See? It's what you might call a hack, but it's the easiest solution I could think of.

Reply:
Thanks!!!! That's super clear :D Thanks for taking the time. Don't be afraid to put lengthy explanations like these in comments within the code, I often do it (e.g. the repetition multimap algorithm thingamagig) and it helps me when looking back on it months later (and no one has complained about it yet :D)